### ISSUE TITLE
Add torch.cat support for torch native sparse tensors. (Need for PyG)

### ISSUE THREAD

**Issue by puririshi98** [pytorch/pytorch#98947]
### 🚀 The feature, motivation and pitch

Up until recently PyG required torch-sparse which provided functionality for sparse tensor math for GNNs. PyG 2.3+ has the goal of dropping torch-sparse and other legacy torch-* packages that PyG used to require. However, for dropping torch-sparse, we are relying on upstream pytorch native sparse tensors. One required functionality of these tensors is the ability to `torch.cat` them together. Currently this is not supported:
This issue tracks the failure w/ reproduction steps: https://github.com/pytorch/pytorch/issues/98861

**Comment by cpuhrsch** [comment]
@puririshi98 - Can we get an exhaustive list of features required for you? Is it just cat or is it foreseeable that there are others? You can explore this by converting to/from dense whenever you encounter a missing operation.

**Comment by puririshi98** [comment]
https://github.com/rusty1s/pytorch_sparse/blob/master/torch_sparse/tensor.py
ideally we would want full support of whatever torch_sparse/tensor.py has. @rusty1s may know what other ops are needed. as of rn I only see failures for torch sum  ([issue](https://github.com/pytorch/pytorch/issues/98796)) and cat

**Comment by puririshi98** [comment]
@rusty1s may know more from his vast knowledge of pyg. i will also post any others that come up for me

**Comment by puririshi98** [comment]
https://github.com/rusty1s/pytorch_sparse/blob/master/torch_sparse/cat.py
i should be able to copy over support for cat from here

**Comment by rusty1s** [comment]
@cpuhrsch I believe `torch.cat` support and support for reductions would be fine. We can also bypass that on our end by building some Python wrapper functions, but ultimately it would be great to have this in PyTorch natively supported.

**Comment by puririshi98** [comment]
https://github.com/pyg-team/pytorch_geometric/pull/7252/files
this PR handles torch.cat for now

**Comment by tvercaut** [comment]
We faced a similar issue with `torch.stack` not supporting CSR matrix. We implemented a workaround here:
https://github.com/cai4cai/torchsparsegradutils/blob/main/torchsparsegradutils/utils/utils.py

Would be great to get this fixed upstream though.

**Comment by tvercaut** [comment]
Yes indeed, the `stack_csr` workaround only addresses this use case but as I understood it from the docs (and trial and error), this is in line with current limitations of batched CSR: 
https://pytorch.org/docs/stable/sparse.html#functionality-overview
